File: qtest-library/src/main/java/io/paulbaker/qtest/LoginToken.java
package io.paulbaker.qtest;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
import lombok.Value;
import okhttp3.Authenticator;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.Route;

import java.io.IOException;
import java.util.Arrays;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Created by Paul N. Baker on 04/13/2018
 */
@Value
@JsonDeserialize(using = LoginToken.LoginTokenDeserializer.class)
public class LoginToken implements Authenticator {

    private String accessToken;
    private String tokenType;
    private String refreshToken;
    private Set<String> scope;
    private String agent;

    @Override
    public Request authenticate(Route route, Response response) throws IOException {
        return null;
    }

    public static class LoginTokenDeserializer extends StdDeserializer<LoginToken> {

        public LoginTokenDeserializer() {
            this(null);
        }

        public LoginTokenDeserializer(Class<?> valueClass) {
            super(valueClass);
        }

        @Override
        public LoginToken deserialize(JsonParser p, DeserializationContext ctxt) throws IOException, JsonProcessingException {
            // This circular logic is strange, but how Jackson works in this context.
            JsonNode treeNode = p.getCodec().readTree(p);
            String accessToken = asNullableText(treeNode.get("access_token"));
            String tokenType = asNullableText(treeNode.get("token_type"));
            String refreshToken = asNullableText(treeNode.get("refresh_token"));
            String scope = Optional.ofNullable(treeNode.get("scope")).map(JsonNode::asText).orElse("");
            Set<String> itemizedScope = Arrays.stream(scope.split("\\s+")).collect(Collectors.toSet());
            String agent = asNullableText(treeNode.get("agent"));
            return new LoginToken(accessToken, tokenType, refreshToken, itemizedScope, agent);
        }

        private String asNullableText(JsonNode jsonNode) {
            if (jsonNode == null) {
                return null;
            }
            return jsonNode.asText(null);
        }
    }
}

File: qtest-library/src/main/kotlin/io/paulbaker/qtest/LoginToken.kt
package io.paulbaker.qtest

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.*
import java.util.function.Supplier


/**
 * @see <a href="https://api.qasymphony.com/#/login/postAccessToken">qTest API</a>
 */
class LoginTokenSupplier(private val site: String, private val username: String, private val password: String) : Supplier<LoginToken> {

    override fun get(): LoginToken {
        val builder = Request.Builder()
        builder.url("https://$site.qtestnet.com/oauth/token")
        builder.post(
                FormBody.Builder()
                        .add("grant_type", "password")
                        .add("username", username)
                        .add("password", password)
                        .build()
        )
        val encoder = Base64.getEncoder()
        builder.addHeader("Authorization", "Basic " + encoder.encodeToString(("$site:").toByteArray()))
        val request = builder.build()

        val client = OkHttpClient.Builder().build()
        val response = client.newCall(request).execute()

        // Serializing to map. I want the scope to be a set of items, not just a massively long string.
        return response.body().use {
            val jacksonObjectMapper = jacksonObjectMapper()
            val body = it!!.string()
            val mapTypeReference = object : TypeReference<Map<String, Any>>() {}
            val jsonMap = jacksonObjectMapper.readValue<Map<String, String?>>(body, mapTypeReference)

            val accessToken = jsonMap["access_token"].asNullableString()
            val tokenType = jsonMap["token_type"].asNullableString()
            val refreshToken = jsonMap["refresh_token"].asNullableString()

            val scope = Regex("\\s+").split(jsonMap["scope"].asNullableString() ?: "").toSet()

            val agent = jsonMap["agent"].asNullableString()
            LoginToken(accessToken, tokenType, refreshToken, scope, agent)
        }
    }

    private fun String?.asNullableString(): String? {
        if (this == "null") {
            return null
        }
        return this
    }
}

/**
 * @see <a href="https://api.qasymphony.com/#/login/postAccessToken">qTest API</a>
 */
data class LoginToken(val accessToken: String?, val tokenType: String?, val refreshToken: String?, val scope: Set<String>, var agent: String?)